{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from datetime import datetime\n",
    "\n",
    "import gget\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import arcadia_pycolor as apc\n",
    "apc.mpl_setup()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_cover_thresh = 50 #min % of query covered for a homolog \n",
    "perident_thresh= 30 #min % percent identity for a homolog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import blast results and get gene names from each ensembl id\n",
    "blast_results_folder = '../inputs/dRE_BlastResults_2023_06_09/'\n",
    "gene_names_path = '..inputs/geneID_Region_andPathogenicLength.csv'\n",
    "\n",
    "try: \n",
    "    genenames = pd.read_csv(gene_names_path)\n",
    "except:\n",
    "    genenames = {}\n",
    "    for i, filename in enumerate(os.listdir(blast_results_folder)):\n",
    "\n",
    "        ensembl_id = filename.replace('.csv','')\n",
    "        geneinfo = gget.info(ensembl_id)\n",
    "        name= geneinfo[\"primary_gene_name\"].values\n",
    "        genenames.update({ensembl_id:name[0]})\n",
    "        print(str(i+1) + \"/\" + str(len(os.listdir(blast_results_folder))) + \" gene names found\")\n",
    "\n",
    "    genenames = pd.DataFrame.from_dict(genenames, orient = \"index\")\n",
    "    genenames.reset_index(inplace=True)\n",
    "    genenames.rename(columns={\"index\":\"Eid\", 0: \"gene\"}, inplace=True)\n",
    "    genenames.to_csv(gene_names_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results_folder = '../inputs/dRE_BlastResults_2023_06_09/'\n",
    "gene_names_path = '../inputs/geneID_Region_andPathogenicLength.csv'\n",
    "genenames = pd.read_csv(gene_names_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter blast results by percent identity and query coverage\n",
    "results_list = []\n",
    "\n",
    "for filename in os.listdir(blast_results_folder):\n",
    "    filepath = blast_results_folder + filename\n",
    "    blast_result = pd.read_csv(filepath)\n",
    "    \n",
    "    ensembl_id = filename.replace('.csv','')\n",
    "    blast_result['gene'] = genenames.loc[genenames[\"GeneId\"].str.match(ensembl_id),\"Locus ID\"].values[0]\n",
    "\n",
    "\n",
    "    blast_result['Query Cover'] = blast_result['Query Cover'].str.rstrip('%').astype('float')\n",
    "    blast_result['Per. Ident'] = blast_result['Per. Ident'].str.rstrip('%').astype('float')                      \n",
    "\n",
    "    results_list.append(blast_result)\n",
    "\n",
    "results_df = pd.concat(results_list)\n",
    "results_df = results_df[\n",
    "            (results_df['Per. Ident'] > perident_thresh) & \n",
    "            (results_df['Query Cover'] > query_cover_thresh)\n",
    "            ].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Description</th>\n",
       "      <th>Scientific Name</th>\n",
       "      <th>Common Name</th>\n",
       "      <th>Taxid</th>\n",
       "      <th>Max Score</th>\n",
       "      <th>Total Score</th>\n",
       "      <th>Query Cover</th>\n",
       "      <th>E value</th>\n",
       "      <th>Per. Ident</th>\n",
       "      <th>Acc. Len</th>\n",
       "      <th>Accession</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>frataxin, mitochondrial isoform 1 preproprotei...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>human</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.000000e-152</td>\n",
       "      <td>100.00</td>\n",
       "      <td>210.0</td>\n",
       "      <td>NP_000135.2</td>\n",
       "      <td>FXN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>FXN [synthetic construct]</td>\n",
       "      <td>synthetic construct</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32630.0</td>\n",
       "      <td>429.0</td>\n",
       "      <td>429.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.000000e-151</td>\n",
       "      <td>99.52</td>\n",
       "      <td>210.0</td>\n",
       "      <td>AKI71332.1</td>\n",
       "      <td>FXN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>FXN [synthetic construct]</td>\n",
       "      <td>synthetic construct</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32630.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>6.000000e-151</td>\n",
       "      <td>99.52</td>\n",
       "      <td>210.0</td>\n",
       "      <td>AKI71333.1</td>\n",
       "      <td>FXN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>frataxin [Homo sapiens]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>human</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.000000e-150</td>\n",
       "      <td>99.05</td>\n",
       "      <td>210.0</td>\n",
       "      <td>AAA98508.1</td>\n",
       "      <td>FXN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>frataxin, mitochondrial isoform X1 [Pan troglo...</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>chimpanzee</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>423.0</td>\n",
       "      <td>423.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4.000000e-149</td>\n",
       "      <td>98.57</td>\n",
       "      <td>210.0</td>\n",
       "      <td>XP_016816418.1</td>\n",
       "      <td>FXN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262857</th>\n",
       "      <td>3648</td>\n",
       "      <td>unnamed protein product [Trichobilharzia regenti]</td>\n",
       "      <td>Trichobilharzia regenti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>157069.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>1.000000e-29</td>\n",
       "      <td>38.50</td>\n",
       "      <td>244.0</td>\n",
       "      <td>VDQ09066.1</td>\n",
       "      <td>GIPC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262858</th>\n",
       "      <td>3655</td>\n",
       "      <td>hypothetical protein TSAR_003650 [Trichomalops...</td>\n",
       "      <td>Trichomalopsis sarcophagae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>543379.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>6.000000e-28</td>\n",
       "      <td>33.94</td>\n",
       "      <td>281.0</td>\n",
       "      <td>OXU17379.1</td>\n",
       "      <td>GIPC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262859</th>\n",
       "      <td>3656</td>\n",
       "      <td>PDZ domain-containing protein GIPC1-like [Fols...</td>\n",
       "      <td>Folsomia candida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>158441.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>3.000000e-27</td>\n",
       "      <td>39.77</td>\n",
       "      <td>180.0</td>\n",
       "      <td>XP_021953350.1</td>\n",
       "      <td>GIPC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262860</th>\n",
       "      <td>3684</td>\n",
       "      <td>unnamed protein product [Heligmosomoides polyg...</td>\n",
       "      <td>Heligmosomoides polygyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6339.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>3.000000e-22</td>\n",
       "      <td>32.61</td>\n",
       "      <td>186.0</td>\n",
       "      <td>VDO29859.1</td>\n",
       "      <td>GIPC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262861</th>\n",
       "      <td>3771</td>\n",
       "      <td>PDZ domain-containing protein GIPC1 [Orchesell...</td>\n",
       "      <td>Orchesella cincta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48709.0</td>\n",
       "      <td>61.2</td>\n",
       "      <td>106.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>7.000000e-07</td>\n",
       "      <td>30.50</td>\n",
       "      <td>183.0</td>\n",
       "      <td>ODM87026.1</td>\n",
       "      <td>GIPC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>262862 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0                                        Description  \\\n",
       "0                0  frataxin, mitochondrial isoform 1 preproprotei...   \n",
       "1                1                          FXN [synthetic construct]   \n",
       "2                2                          FXN [synthetic construct]   \n",
       "3                3                            frataxin [Homo sapiens]   \n",
       "4                4  frataxin, mitochondrial isoform X1 [Pan troglo...   \n",
       "...            ...                                                ...   \n",
       "262857        3648  unnamed protein product [Trichobilharzia regenti]   \n",
       "262858        3655  hypothetical protein TSAR_003650 [Trichomalops...   \n",
       "262859        3656  PDZ domain-containing protein GIPC1-like [Fols...   \n",
       "262860        3684  unnamed protein product [Heligmosomoides polyg...   \n",
       "262861        3771  PDZ domain-containing protein GIPC1 [Orchesell...   \n",
       "\n",
       "                   Scientific Name Common Name     Taxid  Max Score  \\\n",
       "0                     Homo sapiens       human    9606.0      431.0   \n",
       "1              synthetic construct         NaN   32630.0      429.0   \n",
       "2              synthetic construct         NaN   32630.0      427.0   \n",
       "3                     Homo sapiens       human    9606.0      426.0   \n",
       "4                  Pan troglodytes  chimpanzee    9598.0      423.0   \n",
       "...                            ...         ...       ...        ...   \n",
       "262857     Trichobilharzia regenti         NaN  157069.0      124.0   \n",
       "262858  Trichomalopsis sarcophagae         NaN  543379.0      121.0   \n",
       "262859            Folsomia candida         NaN  158441.0      116.0   \n",
       "262860   Heligmosomoides polygyrus         NaN    6339.0      103.0   \n",
       "262861           Orchesella cincta         NaN   48709.0       61.2   \n",
       "\n",
       "        Total Score  Query Cover        E value  Per. Ident  Acc. Len  \\\n",
       "0             431.0        100.0  2.000000e-152      100.00     210.0   \n",
       "1             429.0        100.0  2.000000e-151       99.52     210.0   \n",
       "2             427.0        100.0  6.000000e-151       99.52     210.0   \n",
       "3             426.0        100.0  2.000000e-150       99.05     210.0   \n",
       "4             423.0        100.0  4.000000e-149       98.57     210.0   \n",
       "...             ...          ...            ...         ...       ...   \n",
       "262857        124.0         55.0   1.000000e-29       38.50     244.0   \n",
       "262858        121.0         78.0   6.000000e-28       33.94     281.0   \n",
       "262859        116.0         51.0   3.000000e-27       39.77     180.0   \n",
       "262860        103.0         52.0   3.000000e-22       32.61     186.0   \n",
       "262861        106.0         53.0   7.000000e-07       30.50     183.0   \n",
       "\n",
       "             Accession   gene  \n",
       "0          NP_000135.2    FXN  \n",
       "1           AKI71332.1    FXN  \n",
       "2           AKI71333.1    FXN  \n",
       "3           AAA98508.1    FXN  \n",
       "4       XP_016816418.1    FXN  \n",
       "...                ...    ...  \n",
       "262857      VDQ09066.1  GIPC1  \n",
       "262858      OXU17379.1  GIPC1  \n",
       "262859  XP_021953350.1  GIPC1  \n",
       "262860      VDO29859.1  GIPC1  \n",
       "262861      ODM87026.1  GIPC1  \n",
       "\n",
       "[262862 rows x 13 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_names = pd.read_csv('../inputs/NCBI_taxdump_20230612/names.dmp', delimiter=\"|\", header=None, quotechar='\\t')\n",
    "#fix taxids that got missed\n",
    "tax_names.columns = ['Taxid','Scientific Name','NA1','NA2','NA3']\n",
    "\n",
    "# Filter for errors and extract scientific name from within brackets of the \"description\" column\n",
    "taxids_to_fix = results_df[results_df[\"Taxid\"]==0].copy().reset_index() \n",
    "taxids_to_fix[\"Scientific Name\"] = taxids_to_fix[\"Description\"].str.extract(r'\\[(.*?)\\]')\n",
    "\n",
    "#Merge the filtered DataFrame with the mapping DataFrame\n",
    "merged_df = pd.merge(taxids_to_fix, tax_names[['Scientific Name', 'Taxid']], on='Scientific Name', how='left')\n",
    "\n",
    "#Update the \"Taxid\" column with values from the mapping DataFrame\n",
    "merged_df['Taxid'] = merged_df['Taxid_y'].fillna(merged_df['Taxid_x']).astype(int)\n",
    "\n",
    "# Drop unnecessary columns from the merged DataFrame\n",
    "merged_df = merged_df.drop(['Taxid_x', 'Taxid_y'], axis=1)\n",
    "\n",
    "#create a copy and update Taxid and Scientific Name\n",
    "temp_df = results_df.copy()\n",
    "temp_df.loc[temp_df['Taxid'] == 0, 'Taxid'] = merged_df['Taxid'].values\n",
    "\n",
    "results_df.update(temp_df)\n",
    "\n",
    "#drop those we couldn't fix\n",
    "results_df = results_df[results_df[\"Taxid\"]!=0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "today = datetime.today().strftime('%Y%m%d')\n",
    "\n",
    "#save homology results_df to csv\n",
    "filename = \"../results/homology_results/dREhomologs_\" + today +\".csv\"\n",
    "results_df.to_csv(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#output taxid and homologs for making trees\n",
    "\n",
    "thisdata = results_df[['Taxid','gene']].drop_duplicates()\n",
    "taxid_nhomologs = thisdata[['Taxid']].value_counts().reset_index()\n",
    "\n",
    "filename = \"../results/taxid_nhomologs_forEW\" + today + \".csv\"\n",
    "taxid_nhomologs.to_csv(filename)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "REpilot_blasttest",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
